Shorten Firebase Dynamic Links

unstablebrainiac opened the pull request:
Fixes #708

unstablebrainiac committed:
Shorten share URLs

unstablebrainiac committed:
Add SHORT param

tasomaniac commented on app/src/main/java/org/gdg/frisbee/android/api/FirebaseDynamicLinksHub.java:
```diff
@@ -0,0 +1,22 @@
+import retrofit2.http.Query;
+
+/**
+ * Created by unstablebrainiac on 26/5/17.
```
Please remove the comment. 😄  thanks

tasomaniac commented on app/src/main/java/org/gdg/frisbee/android/api/FirebaseDynamicLinksHub.java:
```diff
@@ -0,0 +1,22 @@
+
+    @FormUrlEncoded
+    @POST("/v1/shortLinks")
+    Call<FirebaseDynamicLinksResponse> getShortenedUrl(@Query("key") String api_key, @Body RequestBody body);
```
Just naming: Can you rename `api_key` to `apiKey`. We try not to use snail case. 

Also what about `shortenUrl` as a method name?

tasomaniac commented on app/src/main/java/org/gdg/frisbee/android/api/FirebaseDynamicLinksHubFactory.java:
```diff
@@ -0,0 +1,34 @@
+ * Created by unstablebrainiac on 26/5/17.
+ */
+
+public final class FirebaseDynamicLinksHubFactory {
```
`Hub` is the name for the api for GDG-X. Do we need `Hub` naming here? I would prefer `FirebaseDynamicLinksFactory`

tasomaniac commented on app/src/main/java/org/gdg/frisbee/android/api/FirebaseDynamicLinksHub.java:
```diff
@@ -0,0 +1,22 @@
+ * Created by unstablebrainiac on 26/5/17.
+ */
+
+public interface FirebaseDynamicLinksHub {
```
What about just `FirebaseDynamicLinks`

tasomaniac commented on app/src/main/java/org/gdg/frisbee/android/api/model/FirebaseDynamicLinksResponse.java:
```diff
@@ -0,0 +1,31 @@
+ */
+
+public class FirebaseDynamicLinksResponse {
+    private String shortLink;
```
Please mark the fields `final`. With this you need to remove the `setter` methods.

tasomaniac commented on app/src/main/java/org/gdg/frisbee/android/onboarding/AppInviteLinkGenerator.java:
```diff
@@ -63,4 +78,30 @@ private String createDeepLink(String gplusId) {
+            @Override
+            public void onSuccess(FirebaseDynamicLinksResponse response) {
+                createChooser(HttpUrl.parse(response.getShortLink()));
+                shareProgressDialog.dismiss();
```
Let's try to decouple the UI concerns and the actual link shortening. 

Now the link creation is asynchronous. Ideally we should have a Listener with a possible method `onDynamicLinkReady(String url)`. Listener's method would be called with the shortened link when it is successful. 

Then the place where we implement this Listener can handle progress show/dismiss. 

What do you think?

unstablebrainiac replied:
Resolved all the other errors but did not understand this very well.

Isn't the Retrofit callback the method that you're talking about? Do you want me to create a separate method and call it from within `onSuccess(FirebaseDynamicLinksResponse response)`? If yes, how do you propose to show the `ProgressDialog` before making the request?

Also, is there anything similar in the codebase that'll probably give me a better idea?

tasomaniac replied:
Unfortunately not. The app doesn't follow any architecture and async handling codes are really old. We didn't have the chance to refactor yet. 

Would you be up to pair on this? We can connect via ScreenHero and do it together.

unstablebrainiac replied:
Sure thing. But Screenhero is not accepting new signups right now so you'll have to send me an invite. My email ID is sajalnarang@gmail.com and I am UnstableBrainiac on Slack.

unstablebrainiac replied:
Hey! Sorry for the delayed response. I've been trying to get Screenhero to work on my system but couldn't. I even re-installed Windows to get it working, but in vain.

I can still contribute though. Let me know what you need me to do.

tasomaniac replied:
I didn't have time to check the latest version of the PR. Sorry. 
Let's schedule some time and talk about it. Can also be a chat or Hangouts video call.

tasomaniac replied:
I will review once more and get back to you.

tasomaniac commented on app/src/main/java/org/gdg/frisbee/android/onboarding/AppInviteLinkGenerator.java:
```diff
@@ -63,4 +78,30 @@ private String createDeepLink(String gplusId) {
 
+    private void shareShortAppInviteLink(String longUrl) {
+        FirebaseDynamicLinksHub firebaseDynamicLinksHub = App.from(activity).getFirebaseDynamicLinksHub();
+        String requestString = "{\"longDynamicLink\": \"" + longUrl + "\",\"suffix\": {\"option\": \"SHORT\"}}";
```
Just like the response, can you create a `Request` class with these fields? 
Then Retrofit and Gson will automatically create the json body for you. This will be much cleaner.

tasomaniac commented on app/build.gradle:
```diff
@@ -66,6 +66,7 @@ android {
 
             buildConfigField "String", "IP_SIMPLE_API_ACCESS_KEY", "\"$local_properties.ip_simple_api_access_key\""
             buildConfigField "String", "ANDROID_SIMPLE_API_ACCESS_KEY", "\"$local_properties.android_simple_api_access_key\""
+            buildConfigField "String", "FIREBASE_WEB_API_KEY", "\"$local_properties.firebase_web_api_key\""
```
Are you sure that you need a new api key? I think `IP_SIMPLE_API_ACCESS_KEY` would work. If you don't have that key, simple create one in Google api console. If you're not sure, feel free to ask us at https://gitter.im/gdg-x/frisbee

Ponderchan replied:
this way may not right. hard to say

tasomaniac reviewed (changes requested):
Thanks a lot for the PR. Great work. 🎉 
I have some small requests and some questions. Let's go over them and merge this. 💯

tasomaniac commented:
Can you also put screen recording or screenshot (with the progress dialog) after you addressed the comments. Thanks.

unstablebrainiac committed:
Align code with project conventions

unstablebrainiac commented:
Here is a [screencast](https://www.youtube.com/watch?v=qRbF7ggTPkI) of the app. I have addressed all the comments. Let me know if there are any more changes to be made.

tasomaniac commented:
Sorry that I still didn't review the code. By the time, Google released a new apis to do this in an easier way. Would you be OK to check that out and apply? I think we can also merge this PR and iterate. What do you think? 

https://firebase.googleblog.com/2017/06/making-dynamic-links-easier.html

unstablebrainiac commented:
You can merge this PR and create a separate issue for the new APIs

tasomaniac commented:
@unstablebrainiac sorry that your PR still not merged. You have problems with checkstyle. Can you look at the report from Travis and fix little problems with the code.

unstablebrainiac committed:
Make style changes to align with repository conventions

unstablebrainiac committed:
Fix indentation issues
